src/metabase/models/field_values.clj

```diff
@@ -138,28 +153,42 @@
 
 (t2/define-before-insert :model/FieldValues
   [{:keys [field_id] :as field-values}]
-  (u/prog1 (merge {:type :full}
-                  field-values)
+  (u/prog1 (merge {:type :full} field-values)
     (assert-valid-human-readable-values field-values)
-    (assert-valid-field-values-type field-values)
-    ;; if inserting a new full fieldvalues, make sure all the advanced field-values of this field is deleted
-    (when (= (:type <>) :full)
+    (assert-valid-type-hash-combo field-values)
+    ;; if inserting a new full fieldvalues, make sure all the advanced field-values of this field are deleted
+    (when (= :full (keyword (:type <>)))
       (clear-advanced-field-values-for-field! field_id))))
 
 (t2/define-before-update :model/FieldValues
   [field-values]
-  (let [{:keys [type values hash_key]} (t2/changes field-values)]
-    (u/prog1 field-values
-      (assert-valid-human-readable-values field-values)
-      (when (or type hash_key)
-        (throw (ex-info (tru "Can't update type or hash_key for a FieldValues.")
-                        {:type        type
-                         :hash_key    hash_key
-                         :status-code 400})))
-      ;; if we're updating the values of a Full FieldValues, delete all Advanced FieldValues of this field
-      (when (and values
-                 (= (:type field-values) :full))
-        (clear-advanced-field-values-for-field! (:field_id field-values))))))
+  (u/prog1 field-values
+    (assert-no-identity-changes field-values)
+    (assert-valid-human-readable-values field-values)
+    ;; if we're updating the values of a Full FieldValues, delete all Advanced FieldValues of this field
+    (when (and (contains? field-values :values) (= :full (keyword (:type field-values))))
```

Comment:
Similar to https://github.com/metabase/metabase/pull/38751/files#r1490642171, does this have a test?

Reply:
The field is non-nullable in the database, so we can't hit this edge case in practice or test it via `update!`. I added this before realising that the update hook gets a fully populated Toucan instance, not the map passed by the caller.

Reply:
> the update hook gets a fully populated Toucan instance

Isn't this change introducing a bug then? `(contains? field-values :values)` will always be `true`, even if we're not updating the `values` field

Reply:
Ah yeah, thanks for spotting the regression. I think this could have hurt performance quite a bit because we keep track of the last accessed time on these records, and would have blown away all the derived fields quite often.

Will add a failing test before fixing.